rules/value.go

```diff
@@ -0,0 +1,333 @@
+	closedCh    chan struct{}
+}
+
+func newRuntimeValue(
+	key string,
+	store kv.Store,
+	unmarshalFn unmarshalFn,
+	versionFn versionFn,
+	cutoverFn cutoverFn,
+	processFn processFn,
+	opts Options,
+) *runtimeValue {
+	rv := &runtimeValue{
+		key:         key,
+		store:       store,
+		unmarshalFn: unmarshalFn,
+		versionFn:   versionFn,
+		cutoverFn:   cutoverFn,
+		processFn:   processFn,
+		opts:        opts,
+		log:         opts.InstrumentOptions().Logger(),
+		nowFn:       opts.ClockOptions().NowFn(),
+		initialized: false,
+		status:      runtimeValueStopped,
+		version:     defaultVersion,
+		closedCh:    make(chan struct{}),
+	}
+	rv.updateFn = rv.Update
+	return rv
+}
+
+func (rv *runtimeValue) Key() string { return rv.key }
+
+func (rv *runtimeValue) Start() error {
+	rv.Lock()
+	if rv.status != runtimeValueStopped {
+		rv.Unlock()
+		return errRuntimeValueNotStopped
+	}
+	watch, err := rv.store.Watch(rv.key)
+	if err != nil {
+		rv.Unlock()
+		return err
+	}
+	rv.watch = watch
+	rv.status = runtimeValueStarted
+	rv.Unlock()
+
+	select {
+	case <-watch.C():
+	case <-time.After(rv.opts.InitWatchTimeout()):
+		err = errInitWatchTimeout
+	}
+
+	if err == nil {
+		err = rv.updateFn(watch.Get())
+	}
+
+	// NB(xichen): we want to start watching updates even though
+	// we may fail to initialize the value temporarily (e.g., during
+	// a network partition) so the value will be updated when the
+	// error condition is resolved
+	rv.updateWg.Add(1)
+	go rv.watchUpdates(watch)
+
+	return err
+}
+
+func (rv *runtimeValue) Update(value kv.Value) error {
+	rv.RLock()
+	if rv.status == runtimeValueClosed {
+		rv.RUnlock()
+		return errRuntimeValueClosed
+	}
+	initialized := rv.initialized
+	rv.RUnlock()
+
+	if !initialized {
+		return rv.initUpdate(value)
+	}
+	return rv.cachedUpdate(value)
+}
+
+func (rv *runtimeValue) Stop() error {
+	rv.Lock()
+	if rv.status != runtimeValueStarted {
+		rv.Unlock()
+		return errRuntimeValueNotStarted
+	}
+	if rv.watch != nil {
+		rv.watch.Close()
+	}
+	rv.watch = nil
+	rv.status = runtimeValueStopping
+	rv.Unlock()
+
+	// NB(xichen): we need to release lock while waiting or
+	// otherwise the stopping goroutine and the updating goroutine
+	// may end up in a deadlock situation
+	rv.updateWg.Wait()
+
+	rv.Lock()
+	defer rv.Unlock()
+
+	// The value may be closed at this point, in which case
+	// we don't override its status
+	if rv.status == runtimeValueStopping {
+		rv.status = runtimeValueStopped
+	}
+	return nil
+}
+
+func (rv *runtimeValue) Close() error {
+	rv.Lock()
+	if rv.status == runtimeValueClosed {
+		rv.Unlock()
+		return errRuntimeValueClosed
+	}
+	if rv.watch != nil {
+		rv.watch.Close()
+	}
+	rv.watch = nil
+	rv.status = runtimeValueClosed
+	rv.Unlock()
+
+	close(rv.closedCh)
+	rv.updateWg.Wait()
+
+	return nil
+}
+
+func (rv *runtimeValue) watchUpdates(watch kv.ValueWatch) {
+	defer rv.updateWg.Done()
+
+	for {
+		select {
+		case _, ok := <-watch.C():
+			if !ok {
+				return
```

Comment:
Hm why is this unexpected? In fact when a namespace is deleted and the corresponding ruleset is expired, we stop watching updates of that ruleset by closing the watch, and may start watching again if that namespace is resurrected later.